### PR TITLE
Filechecker and downloader

### LINKED ISSUE #136: Investigate means of checking whether file is recent or downloaded
PR #76 introduces Decorator, which checks to see whether a file a) exists and b) has been downloaded within a user-specified timeframe; if either condition is not met, the file is then downloaded.

It might be worth investigating NCBI's own version control system and augmenting Decorator's functionality with a means of checking to see whether a file's corresponding database entry has actually had its contents changed. 

This would therefore mean a file is only re-downloaded if any changes have actually happened to it, ideally. A (perhaps extended) timeframe check could still remain just in case.

### PR DESCRIPTION
**Pull request by cemyld**
Resolves #136 and also adds a function for downloading or updating general files from a given url or files with names that correspond to an accession id.

### CODE REVIEW

**Comment by mbonsma** [comment]
Does this replace `check.py` and `test_check.py`? (#76)

**Comment by cemyld** [comment]
It's a more generalized function that also encompasses check.py however instead of checking whether a given amount of time has passed since last update, it checks the date in the accession file's content or uses the last modified header in a url to check whether a file needs updating.

**Comment by mbonsma** [comment]
Ok cool, can you also delete check and test_check in this branch?

File: fetch.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            def print_date(*args, **kwargs):
+               print "Printed by func 'print_date' ", datetime.now()
+
+        >>> print_date(accession_id='NC_000853', months=8)
+
+    Output:
+        NC_000853 File exists and created @ Tue May 17 12:48:14 2016
+        Days passed since last download 0:00:32.932481
+        Printed by func 'print_date' 2016-05-17 12:48:46.932501
+
```
Comment (mbonsma) [inline fetch.py]: Update documentation to (a) reference the original template `check.py` for posterity and (b) include an example for this script.

File: fetch.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    @functools.wraps(test_func)
+    def wrapper(*args, **kwargs):
+        fetch(path, url)
+        return func(*args, **kwargs)
+    return wrapper
+
+"""
+Downloads a file from url to a given path.
+Also shows a progress bar to keep track.
+"""
```
Comment (mbonsma) [inline fetch.py]: Is this docstring for a specific function? If so, put it just below the `def` statement, i.e.

```
def downfile(path, url):
    """
    Downloads a file from url to a given path. 
    Also shows a progress bar to keep track.
    """
```

Comment (mbonsma) [inline fetch.py]: And etc. for the rest of the docstrings.

Comment (cemyld) [inline fetch.py]: Moved the corresponding docstrings to after the function declarations.

File: fetch.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            pbar.update(chunk_size)
+
+"""
+Checks a given path for file and downloads if a url is given or file name
+is an accession id. Also downloads the file if the remote location has a
+more recent version. Gives an IOError if path does not exist and an accession id
+can not be found.
+    Usage:
+        fetch('NC_000853.txt')
+        fetch('DRDatabase.txt', 'http://crispr.i2bc.paris-saclay.fr/crispr/BLAST/DR/DRdatabase')
```
Comment (mbonsma) [inline fetch.py]: Do you run this in python or from the command line? Can you add a bit more detail to the usage part here?

File: fetch.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            return
+        print('File is recent, returning')
+        return
+    """
+    Checks and downloads an accession file from nuccore database if file does
+    not exist or its dates are different from url version.
+    """
+    def gbsync():
+        print('Trying to fetch from Entrez')
+        regex = r'(\w{2}_\d{6}|\w{1}_\d{5})'
```
Comment (mbonsma) [inline fetch.py]: Some accession numbers are longer than 6 numbers after the underscore, and sometimes it's a mix of letters and numbers.

These are all valid accessions, for example:

```
NZ_CP008702
NZ_AHJQ01000001
```

File: fetch.py (hunk 3 above)
Comment (cemyld) [inline fetch.py]: Updated documentation for clarifying usage.

**Comment by mbonsma** [comment]
Looks good to me! So the decorator is still in progress? Either way, I can merge this once the regex is updated.

**Pull request merged**